packages/dev/core/src/Audio/v2/abstractAudioNode.ts

```diff
@@ -0,0 +1,230 @@
+    /**
+     * Creates a new audio node.
+     * @param engine - The audio engine this node will be added to
+     * @param nodeType - The type of audio node
+     */
+    public constructor(engine: AbstractAudioEngine, nodeType: AudioNodeType) {
+        this.engine = engine;
+        engine._addChildNode(this);
+
+        if (nodeType | AudioNodeType.Input) {
+            this._connectedDownstreamNodes = new Array<AbstractAudioNode>();
+        }
+
+        if (nodeType | AudioNodeType.Output) {
+            this._connectedUpstreamNodes = new Array<AbstractAudioNode>();
+        }
+    }
+
+    /**
+     * Releases associated resources.
+     */
+    public dispose(): void {
+        if (this._connectedDownstreamNodes) {
+            for (const node of this._connectedDownstreamNodes) {
+                this._disconnect(node);
+            }
+            this._connectedDownstreamNodes.length = 0;
+        }
+
+        if (this._connectedUpstreamNodes) {
+            for (const node of this._connectedUpstreamNodes) {
+                node._disconnect(this);
+            }
+            this._connectedUpstreamNodes.length = 0;
+        }
+
+        if (this._childNodes) {
+            for (const node of this._childNodes) {
+                node.dispose();
+            }
+            this._childNodes.length = 0;
+        }
+
+        this.parent._removeChildNode(this);
+    }
+
+    public engine: AbstractAudioEngine;
+
+    // If parent is null, node is owned by audio engine.
+    protected _parent: Nullable<IAudioNodeParent> = null;
+
+    public get parent(): IAudioNodeParent {
+        return this._parent ?? this.engine;
+    }
+
+    public setParent(parent: Nullable<IAudioNodeParent>) {
+        if (this._parent === parent) {
+            return;
+        }
+
+        this.parent._removeChildNode(this);
+        this._parent = parent;
+        this.parent._addChildNode(this);
+    }
+
+    protected _childNodes: Nullable<Array<AbstractAudioNode>> = null;
+
+    public _addChildNode(node: AbstractAudioNode): void {
+        if (!this._childNodes) {
+            this._childNodes = new Array<AbstractAudioNode>();
+        } else if (this._childNodes.includes(node)) {
+            return;
+        }
+
+        this._childNodes.push(node);
+    }
+
+    public _removeChildNode(node: AbstractAudioNode): void {
+        if (!this._childNodes) {
+            return;
+        }
+
+        const index = this._childNodes.indexOf(node);
+        if (index < 0) {
+            return;
+        }
+
+        this._childNodes.splice(index, 1);
+    }
+
+    /**
+     * The connected downstream audio nodes.
+     *
+     * Undefined for input nodes.
+     */
+    protected _connectedDownstreamNodes?: Array<AbstractAudioNode>;
+
+    /**
+     * The connected upstream audio nodes.
+     *
+     * Undefined for output nodes.
+     */
+    protected _connectedUpstreamNodes?: Array<AbstractAudioNode>;
```

Comment:
Is it ok for one of these to be `undefined` for the lifetime of the `AbstractAudioNode` object, or is that strange? I was thinking maybe they should be `Nullable` instead, but it would be `null` for the node's entire lifetime if we did that, and we probably shouldn't be adding a property that will never be used, right?

Reply:
We could break this class up into separate classes based on the 3 node types (`Input`, `Output`, and `InputOutput`), but the class hierarchy gets difficult to work with if we do that.

Reply:
It is totally fine for me but the type needs to be changed to `Array<AbstractAudioNode> | undefined`

Reply:
Curious @sebavan, why explicitly add the `| undefined`? It's already implicit from the `?`. I'm mostly curious because I would have wrote this code the same way @docEdub did in this case. I would expect either `?` and no explicit `| undefined`, or no `?` and an explicit `| undefined` (although the two are not semantically the same).

Reply:
I was not aware of this before, but if I'm reading [the `exactOptionalPropertyTypes` docs](https://www.typescriptlang.org/tsconfig/#exactOptionalPropertyTypes)  right, then I **don't** want `Array<AbstractAudioNode> I undefined` because this would allow setting it to `undefined` after the constructor is called. Is that right? So, if I don't want these properties to be set after the constructor is called, then I should leave it as-is, with just the `?`.

Thinking more on this, these properties can be readonly, or private with a protected getter.